### DISCUSSION TITLE
Process question on automatic operations

### DISCUSSION THREAD

**Discussion by 5280Hi** [kachurovskiy/nanoels#148] [Q&A]
Now that I have an H4 implementation that appears to be working as designed, I would like to solicit input on what people have found are the most effective processes to use with the automated operations. It appears to me that getting the soft limits set accurately is the key to obtaining the desired results. For example, if I want to use the automatic turning function to create a 30mm long, 10mm diameter step in a 12mm diameter bar, what steps should I follow? What are the most efficient and accurate procedures to set the stops, and what kid of accuracy should I expect to achieve with the results. I can think of several ways to go about this, but I would like to learn from others who have experience using these automated functions.  Thanks

**Comment by Festivejelly** [comment]
So this isnt super intuitive if you havent done it before so here is the process flow. There is definitely room for improvement from a user experience perspective.

Ideally you would take light cuts, zero X and Y then click automatic turning. Enter a length, depth, pitch and passes then click go.

But its not quite that simple, there are a few hoops to jump through first. We have to set the soft limits. I generally do this when I zero on the Z and X, which you will see in the steps below.

So lets say you wanted to turn something 30mm long and 10mm diameter you would do the following

Put your stock in the lathe. Lets say its 12mm stock and its 40mm long

1. Zero Z: Make a light cut, measure diameter, enter it on the numpad and then press 
![239705610-3059b6ed-0197-4e48-91a7-80a7e1317176](https://github.com/kachurovskiy/nanoels/assets/2408274/615d722d-feed-4485-a724-9bf5da2b6b7a). So if you did a light cut and it measured 11.5mm then you enter that and thats what it will zero the stock to.

2. Set the external limit ![239699957-b18703de-9408-45aa-83e0-812069a3a674](https://github.com/kachurovskiy/nanoels/assets/2408274/218398e9-2eb4-447e-ba91-82c884a77af7)

3. Face stock, to zero Z press ![246621859-32d95cce-d8be-4f8c-8a9c-399d278a2115](https://github.com/kachurovskiy/nanoels/assets/2408274/c64948b3-e799-4446-9cda-4338221aaf2f)

4. Set Z limit ![239699946-9890c846-c7af-4082-a862-39bb0a1ffcd7](https://github.com/kachurovskiy/nanoels/assets/2408274/a35bd867-a32f-4f82-ab4b-5bec17f469b1)

5. Set diameter limit. Move cross slide inwards (X) until it reaches 5mm (10mm divided by 2) then set the limit ![239748452-fd552c49-7f50-4e25-b78f-b0d565c0b96c](https://github.com/kachurovskiy/nanoels/assets/2408274/5fc72e56-ee16-482c-bfc3-ecc445ce71f2)

6. Set the Left max limit. We have already zeroed the Z so we will set the max soft limit to 30 from here by entering 30.00 then pressing 
![239699935-0e3d005e-b917-4389-84b3-76bddbd989a4](https://github.com/kachurovskiy/nanoels/assets/2408274/ad6d833a-21e8-4072-8d4f-e303ff0039ba)

Thats all of our soft limits set, so now we can get onto the actual turning.

7. Turning Outer Diameter: Press ![239707423-00abcae7-0cc3-430d-be56-c8ec22d1d9d0
![239699935-0e3d005e-b917-4389-84b3-76bddbd989a4](https://github.com/kachurovskiy/nanoels/assets/2408274/350c7b8b-db70-4600-843b-15b1396467d8)
](https://github.com/kachurovskiy/nanoels/assets/2408274/59c04135-e3f5-423f-95b5-6fc1b91880e7)
7.1 Set pitch: 0.05mm
7.2 Press ![239707740-c9fb0ef5-94d7-4b42-b1a3-4c85c704e80d](https://github.com/kachurovskiy/nanoels/assets/2408274/97eec6cc-bd6f-4575-979d-c44963c16a94)
7.3 Enter number of passes 
![239707812-3c0e1c27-820a-4d34-ae97-4b362b537e72](https://github.com/kachurovskiy/nanoels/assets/2408274/8e7d35c1-ec21-4420-84ef-c26e456971dc) ![239707815-75db2ae9-97d6-4d39-9fff-e98e889ee84b](https://github.com/kachurovskiy/nanoels/assets/2408274/91c9c954-0dc5-4d2a-b75e-919f92224dbe) confirm with ![239707740-c9fb0ef5-94d7-4b42-b1a3-4c85c704e80d](https://github.com/kachurovskiy/nanoels/assets/2408274/df7a5332-c93c-4ecc-9024-b093eac1b605)
7.4 Turn on the spindle. Operation will proceed fully automatically and the cutter will return to the starting position when done

You will then need to part of the piece by following this guide
https://github.com/kachurovskiy/nanoels/blob/main/h4/README.md#automatic-cut-off

Quick change tool posts help here for quickly swapping between tools. What I do is I measure the offsets of my tool so that its easier to swap between operations.

My parting blade is 1.5mm wide and sits behind the facing too. So when ive faced a part I put the parting tool in, move it up to the face then take a measurement on my DRO and look how far its moved. This give me my offset so I always know where to move my tool after ive done facing.

Having a method to set tool offsets would be handy.

Hope this is useful, please correct any mistakes ive made.

**Reply by 5280Hi** [reply]
Thanks a lot for your reply. Since I wrote the original post, I have been doing a lot of experiments using the automatic operations. For automatic turning, I have essentially been following the process you outline above. This works very well, and most of the time the final dimensions (length and diameter) are + or - 0.07 mm of the set points. If I need to be more accurate, like within 0 .01 mm of my desired dimensions, I do it with multiple auto turning operations. The first time through I set the left stop and inside stop  0.10 mm short of the desired dimension. Then I make 1 or more single pass auto turn operation(s), advancing these 2 stops half the remaining distance with each pass. The last pass is usually a spring pass.

This works very well, but requires a lot of button pushes. It seems to me it would be nice to be able to adjust any stop relative to its current setting and repeat the previous operation with just a few button pushes. I am curious if other people would find this useful as well?

**Reply by Festivejelly** [reply]
The code is all open source. I'll have a look this weekend to see if I can add some options to make it a bit more clear.

Macros might be a starter. You could store that on the device in the config file and it could be a repeat operation. So if you wanted to make the part over and over you would simply just play the macro, re zero, play the macro etc etc.

Im not super familiar with the code but ill have a bash this weekend with Github Copilot :)

**Reply by Festivejelly** [reply]
Also check out https://kachurovskiy.github.io/lathecode/ you can essentially describe what you want to cut in there and send it to the controller. Its a bit limited at the moment but its got a lot of potential there. Im not a fan of button presses. In my opinion anything more than 3 button presses is too much.

I plan on writing a web based GUI for the controller which essentially will mimic the button presses and will allow you to use a web browser for certain operations. I need a bit more time to get this working though.

![image](https://github.com/kachurovskiy/nanoels/assets/2408274/a7838127-80a5-443b-8146-16c8f30bcd5a)

**Reply by Festivejelly** [reply]
What Maxim has created is awesome so im sure any help any of us can give him to enhance the experience will be welcomed.
